<a href="https://colab.research.google.com/github/mtello5/IS170_Machine_Learning/blob/Association-Rules/AssociationRulesHW5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **HW5 Association Rules | Matthew Tello**

In [ ]:
#set up the libraries
# we use apriori for SUPPORT and association_rules for CONFIDENCE, LIFT and CONVICTION
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder


In [ ]:
#Loading the dataset
dataset=pd.read_csv('/content/G1list.csv')
dataset.head()

Groceries
0          MILK, BREAD, BISCUIT
1  BREAD, MILK, BISCUIT, CEREAL
2          BREAD, BANANAS, EGGS
3       JAM, HONEY, BREAD, MILK
4       HONEY, BANANAS, BISCUIT

In [ ]:
#Since it's missing a column header, so we create new column header name
df=pd.read_csv('/content/G1list.csv', sep=',')
df.head()

Groceries
0          MILK, BREAD, BISCUIT
1  BREAD, MILK, BISCUIT, CEREAL
2          BREAD, BANANAS, EGGS
3       JAM, HONEY, BREAD, MILK
4       HONEY, BANANAS, BISCUIT

In [ ]:
#data size
df.shape

(29, 1)

In [ ]:
#we are turning the pandas dataframe into a list of lists
data=list(df["Groceries"].apply(lambda x:x.split(",")))
data

[['MILK', ' BREAD', ' BISCUIT'],
 ['BREAD', ' MILK', ' BISCUIT', ' CEREAL'],
 ['BREAD', ' BANANAS', ' EGGS'],
 ['JAM', ' HONEY', ' BREAD', ' MILK'],
 ['HONEY', ' BANANAS', ' BISCUIT'],
 ['BREAD', ' BANANAS', ' EGGS'],
 ['HONEY', ' BANANAS', ' CEREAL'],
 ['HONEY', ' BREAD', ' BANANAS', ' BISCUIT'],
 ['JAM', ' HONEY', ' BREAD', ' BANANAS'],
 ['BREAD', ' MILK'],
 ['COFFEE', ' COCK', ' BISCUIT', ' CEREAL'],
 ['COFFEE', ' COCK', ' BISCUIT', ' CEREAL'],
 ['COFFEE', ' SUGAR', ' EGGS'],
 ['BREAD', ' COFFEE', ' COCK'],
 ['BREAD', ' SUGAR', ' BISCUIT'],
 ['COFFEE', ' SUGAR', ' CEREAL'],
 ['BREAD', ' SUGAR', ' EGGS'],
 ['BREAD', ' COFFEE', ' SUGAR'],
 ['BREAD', ' COFFEE', ' SUGAR'],
 ['BANANAS', ' MILK', ' COFFEE', ' CEREAL'],
 ['WATER', ' BREAD', ' BISCUIT'],
 ['BREAD', ' HONEY', ' BREAD', ' COFFEE'],
 ['COFFEE', ' BREAD', ' BUTTER', ' BREAD'],
 ['HONEY', ' MILK', ' COFFEE', ' BISCUIT'],
 ['BREAD', ' COFFEE', ' BISCUIT', ' COOKIES'],
 ['HONEY', ' BANANAS', ' BREAD', ' COFFEE'],
 ['BREAD', ' BANA

In [ ]:
#Let's transform the list with one-hot encoding
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)
df=df.replace(True,1)
df=df.replace(False,0)
df

BANANAS   BISCUIT   BREAD   BUTTER   CEREAL   COCK   COFFEE   COOKIES  \
0          0         1       1        0        0      0        0         0   
1          0         1       0        0        1      0        0         0   
2          1         0       0        0        0      0        0         0   
3          0         0       1        0        0      0        0         0   
4          1         1       0        0        0      0        0         0   
5          1         0       0        0        0      0        0         0   
6          1         0       0        0        1      0        0         0   
7          1         1       1        0        0      0        0         0   
8          1         0       1        0        0      0        0         0   
9          0         0       0        0        0      0        0         0   
10         0         1       0        0        1      1        0         0   
11         0         1       0        0        1      1        0         0   
12         0         0       0        0        0      0        0         0   
13         0         0       0        0        0      1        1         0   
14         0         1       0        0        0      0        0         0   
15         0         0       0        0        1      0        0         0   
16         0         0       0        0        0      0        0         0   
17         0         0       0        0        0      0        1         0   
18         0         0       0        0        0      0        1         0   
19         0         0       0        0        1      0        1         0   
20         0         1       1        0        0      0        0         0   
21         0         0       1        0        0      0        1         0   
22         0         0       1        1        0      0        0         0   
23         0         1       0        0        0      0        1         0   
24         0         1       0        0        0      0        1         1   
25         1         0       1        0        0      0        1         0   
26         1         1       0        0        0      0        0         0   
27         1         0       1        0        0      0        1         0   
28         0         1       0        0        0      0        1         0   

     EGGS   HONEY   MILK   SUGAR  BANANAS  BREAD  COFFEE  HONEY  JAM  MILK  \
0       0       0      0       0        0      0       0      0    0     1   
1       0       0      1       0        0      1       0      0    0     0   
2       1       0      0       0        0      1       0      0    0     0   
3       0       1      1       0        0      0       0      0    1     0   
4       0       0      0       0        0      0       0      1    0     0   
5       1       0      0       0        0      1       0      0    0     0   
6       0       0      0       0        0      0       0      1    0     0   
7       0       0      0       0        0      0       0      1    0     0   
8       0       1      0       0        0      0       0      0    1     0   
9       0       0      1       0        0      1       0      0    0     0   
10      0       0      0       0        0      0       1      0    0     0   
11      0       0      0       0        0      0       1      0    0     0   
12      1       0      0       1        0      0       1      0    0     0   
13      0       0      0       0        0      1       0      0    0     0   
14      0       0      0       1        0      1       0      0    0     0   
15      0       0      0       1        0      0       1      0    0     0   
16      1       0      0       1        0      1       0      0    0     0   
17      0       0      0       1        0      1       0      0    0     0   
18      0       0      0       1        0      1       0      0    0     0   
19      0       0      1       0        1      0       0      0    0     0   
20      0       0      0  

In [ ]:
#Set a threshold value for the support value and calculate the support value
df=apriori(df,min_support=0.2, use_colnames=True)
df

support          itemsets
0  0.310345        ( BANANAS)
1  0.413793        ( BISCUIT)
2  0.310345          ( BREAD)
3  0.206897         ( CEREAL)
4  0.344828         ( COFFEE)
5  0.206897           ( MILK)
6  0.206897          ( SUGAR)
7  0.448276           (BREAD)
8  0.206897  ( COFFEE, BREAD)

SUPPORT tells us how frequent each of the items are appearing in our dataset. In this example, we know bananas appears the most with SUPPORT measure of .31. Further, we know bread is the most likely to be seen in a transaction with SUPPORT value .44

In [ ]:
#Here is the CONFIDENCE (Try different min_threshold value to explore this more)
df_ar=association_rules(df,metric="confidence", min_threshold=0.3)
df_ar

antecedents consequents  antecedent support  consequent support   support  \
0   ( COFFEE)     (BREAD)            0.344828            0.448276  0.206897   
1     (BREAD)   ( COFFEE)            0.448276            0.344828  0.206897   

   confidence      lift  leverage  conviction  
0    0.600000  1.338462  0.052319    1.379310  
1    0.461538  1.338462  0.052319    1.216749

We can see (Coffee -> Bread) show CONFIDENCE 0.6 which means the two items are bought together 60% of the time. Additionally, we can see (Bread -> Coffee) shows CONFIDENCE 0.46 meaning when someone buys Bread 46% of the time they buy coffee with it together.

In [ ]:
#Here is LIFT (try different min_threshold value to explore more) 
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents consequents  antecedent support  consequent support   support  \
0   ( COFFEE)     (BREAD)            0.344828            0.448276  0.206897   
1     (BREAD)   ( COFFEE)            0.448276            0.344828  0.206897   

   confidence      lift  leverage  conviction  
0    0.600000  1.338462  0.052319    1.379310  
1    0.461538  1.338462  0.052319    1.216749

The first row (coffee & bread) show LIFT 1.33 which is only slight positive correlation between the two. The next row, (Bread -> Coffee) show LIFT 1.33 which is the same because A->C & C->A there is no antecedent or consequent. 

In [ ]:
#Let's view our interpretation values using the Association rule function
df_ar=association_rules(df,metric="conviction", min_threshold=0.6)
df_ar

antecedents consequents  antecedent support  consequent support   support  \
0   ( COFFEE)     (BREAD)            0.344828            0.448276  0.206897   
1     (BREAD)   ( COFFEE)            0.448276            0.344828  0.206897   

   confidence      lift  leverage  conviction  
0    0.600000  1.338462  0.052319    1.379310  
1    0.461538  1.338462  0.052319    1.216749

CONVICTION is same as LIFT except CONVICTION also gives the directed measure. The first row (Coffee -> Bread) CONVICTION shows 1.37 which is pretty strong relationship, but the second row (Bread -> Coffee) CONVICTION is 1.21 which is also strong relationship but not as strong. The higher than 1 the CONVICTION value the higher the relationship. The antecedent in this case Coffee has a stronger relationship to Bread than Bread to Coffee. 

Let's look at another dataset and example

In [ ]:
#Loading the dataset
dataset3=pd.read_csv('/content/BreakfastData.csv')
dataset3.head()

eggs, bacon, toast
0  eggs, bacon, sausage
1   eggs, bacon, coffee
2   eggs, coffee, toast
3  bacon, coffee, toast
4  eggs, sausage, toast

In [ ]:
#Since it's missing a column header, so we create new column header name
df3=pd.read_csv('/content/BreakfastData.csv', names=['products'])
df3.head() 

products
0    eggs, bacon, toast
1  eggs, bacon, sausage
2   eggs, bacon, coffee
3   eggs, coffee, toast
4  bacon, coffee, toast

In [ ]:
#data size
df3.shape

(11, 1)

In [ ]:
#we are turning the pandas dataframe into a list of lists
data3=list(df3["products"].apply(lambda x:x.split(",")))
data3

[['eggs', ' bacon', ' toast'],
 ['eggs', ' bacon', ' sausage'],
 ['eggs', ' bacon', ' coffee'],
 ['eggs', ' coffee', ' toast'],
 ['bacon', ' coffee', ' toast'],
 ['eggs', ' sausage', ' toast'],
 ['sausage', ' bacon', ' toast'],
 ['sausage', ' bacon', ' coffee'],
 ['sausage', ' coffee', ' toast'],
 ['eggs', ' bacon', ' sausage', ' coffee'],
 ['bacon', ' coffee', ' toast', ' sausage']]

In [ ]:
#Let's transform the list with one-hot encoding (make sure you understand the one-hot encoding concept first)
#Each column represents an item and each row represent a record or a transaction. 
#Initially, if that particular item occurs then it's labeled TRUE, if not then FALSE.
#After that, TRUE is converted to 1 and FALSE to 0 because computer loves numeric values.

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data3).transform(data3)
df3=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df3=df3.replace(False,0)
df3=df3.replace(True,1)

df3

bacon   coffee   sausage   toast  bacon  eggs  sausage
0        1        0         0       1      0     1        0
1        1        0         1       0      0     1        0
2        1        1         0       0      0     1        0
3        0        1         0       1      0     1        0
4        0        1         0       1      1     0        0
5        0        0         1       1      0     1        0
6        1        0         0       1      0     0        1
7        1        1         0       0      0     0        1
8        0        1         0       1      0     0        1
9        1        1         1       0      0     1        0
10       0        1         1       1      1     0        0

In [ ]:
#We'll try 0.2 for the SUPPORT threshold value (but you can set it to any value you want).
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df3=apriori(df3,min_support=0.2,use_colnames=True)
df3

support           itemsets
0   0.545455           ( bacon)
1   0.636364          ( coffee)
2   0.363636         ( sausage)
3   0.636364           ( toast)
4   0.545455             (eggs)
5   0.272727          (sausage)
6   0.272727  ( coffee,  bacon)
7   0.363636     (eggs,  bacon)
8   0.363636  ( coffee,  toast)
9   0.272727    ( coffee, eggs)
10  0.272727   (eggs,  sausage)
11  0.272727     (eggs,  toast)

Bacon we can see has SUPPORT value .545 which if we did it manually means it appears in 5/11 transactions and overall was most prominent in over half. Additionally we can see some itemsets with more than one item also showing SUPPORT. On the other side, we can see lower SUPPORT level mean they itemset is less likely to appear in the transaction.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df3_ar=association_rules(df3,metric="lift", min_threshold=1)
df3_ar

antecedents consequents  antecedent support  consequent support   support  \
0      (eggs)    ( bacon)            0.545455            0.545455  0.363636   
1    ( bacon)      (eggs)            0.545455            0.545455  0.363636   
2      (eggs)  ( sausage)            0.545455            0.363636  0.272727   
3  ( sausage)      (eggs)            0.363636            0.545455  0.272727   

   confidence      lift  leverage  conviction  
0    0.666667  1.222222  0.066116    1.363636  
1    0.666667  1.222222  0.066116    1.363636  
2    0.500000  1.375000  0.074380    1.272727  
3    0.750000  1.375000  0.074380    1.818182

Using LIFT to determine values of relationship we can see relationship of items bought seperately from each other. We can see that they have a postive correlation with LIFT value of 1.222. We can also see eggs and sausage with a LIFT value of 1.375 will a stronger postive correlation.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df3_ar=association_rules(df3,metric="confidence", min_threshold=0.6)
df3_ar

antecedents consequents  antecedent support  consequent support   support  \
0      (eggs)    ( bacon)            0.545455            0.545455  0.363636   
1    ( bacon)      (eggs)            0.545455            0.545455  0.363636   
2  ( sausage)      (eggs)            0.363636            0.545455  0.272727   

   confidence      lift  leverage  conviction  
0    0.666667  1.222222  0.066116    1.363636  
1    0.666667  1.222222  0.066116    1.363636  
2    0.750000  1.375000  0.074380    1.818182

CONFIDENCE is used to tell us probability the antecedant is purchased in the same transcation as the consequent. In this case, if a customer buys eggs there is a 66.7% chance they buy bacon also. Additionally, If someone buys sausage with CONFIDENCE value of .75 there is a 75% chance they buy eggs with it also.

In [ ]:
#Let's view our interpretation values using the Association rule function
df3_ar=association_rules(df3,metric="conviction", min_threshold=0.6)
df3_ar

antecedents consequents  antecedent support  consequent support   support  \
0    ( coffee)    ( bacon)            0.636364            0.545455  0.272727   
1     ( bacon)   ( coffee)            0.545455            0.636364  0.272727   
2       (eggs)    ( bacon)            0.545455            0.545455  0.363636   
3     ( bacon)      (eggs)            0.545455            0.545455  0.363636   
4    ( coffee)    ( toast)            0.636364            0.636364  0.363636   
5     ( toast)   ( coffee)            0.636364            0.636364  0.363636   
6    ( coffee)      (eggs)            0.636364            0.545455  0.272727   
7       (eggs)   ( coffee)            0.545455            0.636364  0.272727   
8       (eggs)  ( sausage)            0.545455            0.363636  0.272727   
9   ( sausage)      (eggs)            0.363636            0.545455  0.272727   
10      (eggs)    ( toast)            0.545455            0.636364  0.272727   
11    ( toast)      (eggs)            0.636364            0.545455  0.272727   

    confidence      lift  leverage  conviction  
0     0.428571  0.785714 -0.074380    0.795455  
1     0.500000  0.785714 -0.074380    0.727273  
2     0.666667  1.222222  0.066116    1.363636  
3     0.666667  1.222222  0.066116    1.363636  
4     0.571429  0.897959 -0.041322    0.848485  
5     0.571429  0.897959 -0.041322    0.848485  
6     0.428571  0.785714 -0.074380    0.795455  
7     0.500000  0.785714 -0.074380    0.727273  
8     0.500000  1.375000  0.074380    1.272727  
9     0.750000  1.375000  0.074380    1.818182  
10    0.500000  0.785714 -0.074380    0.727273  
11    0.428571  0.785714 -0.074380    0.795455

Lastly, CONVICTION lets us know how probable it A appears without C. So by analyzing our list of CONVICTION values we see (Eggs->Bacon) and (Sausage->Eggs) have values above 1 meaning they have a strong relationship. All  the other values we see in the table show CONVICTION values >1 meaning they have negative relationships. This means they antecedent makes the consequent less likely to be in the transaction (coffee -> bacon)

**What is Machine Learning:**


Machine Learning is how computers can make autonomous judgements utilizing large amounts of data and algorithms to identify patterns and relationships in the data. Machine leanring enables the machine to then make its own assumptions and decisions.